src/DynamicRupture/FrictionLaws/FrictionSolver.cpp

```diff
@@ -0,0 +1,209 @@
+}
+
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp declare target
+#endif // ACL_DEVICE_OFFLOAD
+void FrictionSolver::precomputeStressFromQInterpolated(FaultStresses& faultStresses,
+                                                       unsigned int ltsFace) {
+
+  static_assert(tensor::QInterpolated::Shape[0] == tensor::resample::Shape[0],
+                "Different number of quadrature points?");
+
+  // this initialization of the kernel could be moved to the initializer,
+  // since all inputs outside the j-loop are time independent
+  // set inputParam could be extendent for this
+  // the kernel then could be a class attribute (but be careful of race conditions since this is
+  // computed in parallel!!)
+
+  auto etaP = impAndEta[ltsFace].etaP;
+  auto etaS = impAndEta[ltsFace].etaS;
+  auto invZp = impAndEta[ltsFace].invZp;
+  auto invZs = impAndEta[ltsFace].invZs;
+  auto invZpNeig = impAndEta[ltsFace].invZpNeig;
+  auto invZsNeig = impAndEta[ltsFace].invZsNeig;
+
+  using QInterpolatedShapeT =
+      real(*)[CONVERGENCE_ORDER][misc::numQuantities][misc::numPaddedPoints];
+  auto* qIPlus = (reinterpret_cast<QInterpolatedShapeT>(qInterpolatedPlus))[ltsFace];
+  auto* qIMinus = (reinterpret_cast<QInterpolatedShapeT>(qInterpolatedMinus))[ltsFace];
+
+  for (unsigned o = 0; o < CONVERGENCE_ORDER; ++o) {
+
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp loop bind(parallel)
+#endif // ACL_DEVICE_OFFLOAD
+    for (unsigned i = 0; i < misc::numPaddedPoints; ++i) {
+      faultStresses.normalStress[o][i] =
+          etaP * (qIMinus[o][6][i] - qIPlus[o][6][i] + qIPlus[o][0][i] * invZp +
+                  qIMinus[o][0][i] * invZpNeig);
+
+      faultStresses.traction1[o][i] =
+          etaS * (qIMinus[o][7][i] - qIPlus[o][7][i] + qIPlus[o][3][i] * invZs +
+                  qIMinus[o][3][i] * invZsNeig);
+
+      faultStresses.traction2[o][i] =
+          etaS * (qIMinus[o][8][i] - qIPlus[o][8][i] + qIPlus[o][5][i] * invZs +
+                  qIMinus[o][5][i] * invZsNeig);
+    }
+  }
+}
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp end declare target
+#endif // ACL_DEVICE_OFFLOAD
+
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp declare target
+#endif // ACL_DEVICE_OFFLOAD
+void FrictionSolver::postcomputeImposedStateFromNewStress(const FaultStresses& faultStresses,
+                                                          const TractionResults& tractionResults,
+                                                          double timeWeights[CONVERGENCE_ORDER],
+                                                          unsigned int ltsFace) {
+  // this initialization of the kernel could be moved to the initializer
+  // set inputParam could be extendent for this (or create own function)
+  // the kernel then could be a class attribute and following values are only set once
+  //(but be careful of race conditions since this is computed in parallel for each face!!)
+
+  // set imposed state to zero
+#ifdef ACL_DEVICE_OFFLOAD
+#pragma omp loop bind(parallel)
```

Comment:
Also here: do we need an else for cpu only excution?

Reply:
auto-vectorization is done by a compiler